swupd/hash.go

```diff
@@ -34,16 +35,26 @@ var AllZeroHash = "0000000000000000000000000000000000000000000000000000000000000
 // Hashes is a global map of indices to hashes
 var Hashes = []*string{&AllZeroHash}
 var invHash = map[string]Hashval{AllZeroHash: 0}
+var rwMutex sync.RWMutex
 
 // internHash adds only new hashes to the Hashes slice and returns the index at
 // which they are located
 func internHash(hash string) Hashval {
+	rwMutex.RLock()
 	if key, ok := invHash[hash]; ok {
+		rwMutex.RUnlock()
+		return key
+	}
+	rwMutex.RUnlock()
+	rwMutex.Lock()
+	if key, ok := invHash[hash]; ok {
```

Comment:
Wait, isn't this the same check as above but using a `Lock` instead of an `RLock`?

Reply:
Yes

Reply:
Adding a comment to explain why we're doing this twice 'cause it is actually confusing..

Reply:
And it is essential that we do this check a second time.

We could go for lower performance and just Lock() everywhere, rather than using an RWLock to avoid the second check, but it would be a poor tradeoff unless we think that we are only going to look up keys once.